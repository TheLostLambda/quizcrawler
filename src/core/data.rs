use crate::core::logic;

// Just drafting ideas here
#[derive(Debug, Clone)]
pub struct Section {
    name: String,
    questions: Vec<Question>,
    children: Vec<Section>,
}

impl Section {
    // Gah, this is kinda pointless right now...
    pub fn new(name: String, children: Vec<Section>, questions: Vec<Question>) -> Section {
        Section {
            name,
            children,
            questions,
        }
    }
    // So are these. If version 1.0 doesn't make these more meaningful, just
    // make the struct fields public
    pub fn name(&self) -> &str {
        &self.name
    }
    pub fn children(&self) -> &Vec<Section> {
        &self.children
    }
    pub fn is_parent(&self) -> bool {
        self.children().len() > 0
    }
    pub fn questions(&self) -> &Vec<Question> {
        &self.questions
    }
    // Write a test function for this!
    pub fn child_at_path(&self, path: &[&str]) -> Option<&Section> {
        let mut current = self;
        for name in path {
            // What happens when there are sections with the same name?
            current = current
                .children()
                .iter()
                .filter(|c| &c.name() == name)
                .next()?;
        }
        Some(current)
    }
}
// Is it possible to match an identical regex group a second time?
// If so, I should write a regex that matches *'s followed by whitespace and
// should continue until the same number of *'s are encountered again (sibling
// section) or the end of the file is reached. I think I should be able to do
// that with groups.

// Two passes follow, first the body of the section is parsed into questions
// (this may actually involve several subpasses, one for each question type)
// then the body is scanned for children and the process repeats.

// I need to find a way to not double scan the questions within the children â€”
// unless this becomes desirable. Perhaps the children pass happens first and
// those matches are snipped from the string before questions are scanned
// for. It's also possible just to build a second, negated regex that matches
// things that aren't children and scrapes questions from those.

#[derive(Debug, Clone)]
pub enum Strictness {
    Exact,
    Trimmed,
    Caseless,
}

#[derive(Debug, Clone)]
pub struct Question {
    data: QuestionVariant,
    comp_level: Strictness,
    mastery: u8, // (0-5 or 0-10?)
    seen: u32,
    correct: u32,
}

/// Question Enum
#[derive(Debug, Clone)]
pub enum QuestionVariant {
    Term(Term),
    List(List),
    Bullet(Bullet),
    _Equation(),
}

/// Flash Cards
#[derive(Debug, Clone)]
pub struct Term {
    term: String,
    definition: String,
    inverted: bool,
}

// Not a fan of this section having children, but I'll allow it for now
// Also, term? item? body? Pick one.
#[derive(Debug, Clone)]
pub struct List {
    order: u32,
    item: String,
    details: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct Bullet {
    body: String,
}

impl Term {
    pub fn new(term: String, definition: String) -> Question {
        Question::new(QuestionVariant::Term(Self {
            term,
            definition,
            inverted: false,
        }))
    }

    pub fn flip(&mut self) {
        self.inverted = !self.inverted;
    }
}

impl List {
    // Slice of strings here? &[] not Vec?
    pub fn new(order: u32, item: String, details: Vec<String>) -> Question {
        Question::new(QuestionVariant::List(Self {
            order,
            item,
            details,
        }))
    }
}

impl Bullet {
    pub fn new(body: String) -> Question {
        Question::new(QuestionVariant::Bullet(Self { body }))
    }
}

impl Question {
    pub fn new(data: QuestionVariant) -> Question {
        Question {
            data,
            comp_level: Strictness::Trimmed,
            mastery: 0, // (0-5 or 0-10?)
            seen: 0,
            correct: 0,
        }
    }

    pub fn ask(&self) -> &str {
        match &self.data {
            QuestionVariant::Term(t) => {
                if t.inverted {
                    &t.definition
                } else {
                    &t.term
                }
            }
            _ => "",
        }
    }

    pub fn peek(&self) -> &str {
        match &self.data {
            QuestionVariant::Term(t) => {
                if t.inverted {
                    &t.term
                } else {
                    &t.definition
                }
            }
            _ => "",
        }
    }

    pub fn answer(&mut self, ans: &str) -> (bool, &str) {
        let right_ans = self.peek().to_owned();
        let correct = logic::check_answer(ans, &right_ans, &self.comp_level);
        self.seen += 1;
        if correct {
            self.correct += 1;
        }
        (correct, self.peek())
    }

    // I'm not a massive fan of this...
    pub fn inner(&mut self) -> &QuestionVariant {
        &self.data
    }

    pub fn _set_comp_level(&mut self, cl: Strictness) {
        self.comp_level = cl;
    }

    pub fn _get_comp_level(&self) -> &Strictness {
        &self.comp_level
    }
    // Replace this and times_correct with a single "mastery" value
    pub fn _times_seen(&self) -> u32 {
        self.seen
    }

    pub fn _times_correct(&self) -> u32 {
        self.correct
    }

    pub fn override_correct(&mut self) {
        self.correct += 1;
        if self.correct > self.seen {
            self.correct = self.seen;
        }
    }
}

impl PartialEq for Question {
    fn eq(&self, other: &Self) -> bool {
        match (&self.data, &other.data) {
            (QuestionVariant::Term(s), QuestionVariant::Term(o)) => s.term == o.term,
            _ => false,
        }
    }
}
