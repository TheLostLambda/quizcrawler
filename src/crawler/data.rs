use crate::crawler::util::*;
use serde_derive::Deserialize;

/// Flash Cards
#[derive(Debug)]
pub struct Flash {
    term: String,
    definition: String,
}

impl Flash {
    pub fn new(t: &str, d: &str) -> Flash {
        Flash {
            term: t.to_owned(),
            definition: d.to_owned(),
        }
    }
}

/// Generic Result for using `?`
type BoxResult<T> = Result<T,Box<dyn std::error::Error>>;

/// This struct holds all of the configuration data that is parsed from the TOML
#[derive(Debug, Deserialize)]
pub struct Config {
    pub title: String,
    pub exts: Vec<String>,
    pub flash: FlashConfig,
}

impl Config {
    pub fn from_file(file_name: &str) -> BoxResult<Config> {
        let toml_str = read_file_as_string(file_name)?;
        Ok(toml::from_str(&toml_str)?)
    }
}
        
/// This struct holds the regex components needed to extract flashcards
#[derive(Debug, Deserialize)]
pub struct FlashConfig {
    pub leader: String,
    pub term: String,
    pub separator: String,
    pub definition: String,
    pub terminator: String,
}
