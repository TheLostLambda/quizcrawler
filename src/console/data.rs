use crate::core::{
    data::{Section, Path},
    quiz::{Dispatcher, MultipleChoice, QuizProgress, QuizRef, DSettings},
};
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use std::collections::HashMap;

// Trim back things that don't need to be public

#[derive(Default)]
pub struct QCSettings {}

pub struct Quizcrawler {
    pub tree: Section,
    pub settings: QCSettings,
    pub state_stack: Vec<State>,
}

#[derive(Default)]
pub struct TreeState {
    pub path: Path,
    pub selection_history: HashMap<Path, usize>,
}

impl TreeState {
    pub fn get_selected_mut(&mut self) -> &mut usize {
        self.selection_history.entry(self.path.clone()).or_insert(0)
    }

    pub fn get_selected(&self) -> usize {
        *self.selection_history.get(&self.path).unwrap_or(&0)
    }
}

#[derive(Clone)]
pub struct QuestionState {
    pub quiz: QuizRef,
    pub progress: QuizProgress,
}

// FIXME: This could use some more thought
pub enum State {
    TreeView(TreeState),
    Dispatch(Dispatcher),
    AskQuestion(QuestionState),
    AnswerQuestion(QuestionState, (bool, String)), // FIXME: Ew
}

impl Quizcrawler {
    pub fn new(settings: QCSettings, tree: Section) -> Self {
        Self {
            tree,
            settings,
            state_stack: vec![State::TreeView(TreeState::default())],
        }
    }

    pub fn handle_key(&mut self, key: KeyEvent) {
        match self.state_stack.last_mut() {
            Some(State::TreeView(state)) => {
                // FIXME: Maybe move this to a function?
                let node = self.tree.child_at_path(&state.path).unwrap(); // FIXME: Unwrap
                let child_names: Vec<_> = node.children.iter().map(|x| &x.name).collect();
                let limit = child_names.len() - 1;
                let current = state.get_selected();
                // This gives me an entry, a mutable reference which is updated in the match
                let selector = state.get_selected_mut();
                match key.code {
                    KeyCode::Char('q') => {
                        self.state_stack.pop();
                    }
                    KeyCode::Up if current > 0 => *selector -= 1,
                    KeyCode::Down if current < limit => *selector += 1,
                    KeyCode::Right if node.children[current].is_parent() => {
                        state.path.push(child_names[current].to_owned())
                    }
                    // FIXME: This feels a tad muddled
                    KeyCode::Char(' ') => {
                        let mut path = state.path.clone();
                        path.push(child_names[current].to_owned());
                        // FIXME: unwrap is likely a bad idea here
                        let section = self
                            .tree
                            .child_at_path(&path)
                            .unwrap();
                        let settings = DSettings {
                            recursive: key.modifiers.contains(KeyModifiers::CONTROL),
                            ..Default::default()
                        };
                        let mut dispatcher = Dispatcher::new(settings, section);
                        dispatcher.register_quiz(MultipleChoice::default());
                        self.state_stack.push(State::Dispatch(dispatcher))
                    }
                    KeyCode::Left => {
                        state.path.pop();
                    }
                    _ => {}
                }
            }
            Some(State::AskQuestion(state)) => {
                if !state.quiz.borrow().get_choices().is_empty() {
                    match key.code {
                        KeyCode::Char('q') => {
                            // FIXME: This double-pop is hacky and is needed to get past the dispatcher
                            // Write a rewind function to pop the stack back until some condition is met,
                            // like reaching something isn't Dispatch. Maybe the states could have a
                            // transient flag and all of those are popped off.
                            self.state_stack.pop();
                            self.state_stack.pop();
                        }
                        KeyCode::Char('h') => state.quiz.borrow_mut().get_hint(),
                        KeyCode::Char(c) => {
                            let result = state.quiz.borrow_mut().answer(&c.to_string());
                            if let Some(result) = result {
                                let state = state.clone();
                                self.state_stack.pop();
                                self.state_stack.push(State::AnswerQuestion(state, result))
                            }
                        }
                        _ => {}
                    }
                }
            }
            Some(State::AnswerQuestion(state, (correct, _))) => match key.code {
                KeyCode::Char('q') => {
                    // FIXME: Hacky double-pop
                    self.state_stack.pop();
                    self.state_stack.pop();
                }
                KeyCode::Char('o') if !*correct => {
                    state.quiz.borrow_mut().i_was_right();
                    *correct = true;
                }
                KeyCode::Char(' ') => {
                    self.state_stack.pop();
                }
                _ => {}
            },
            _ => {}
        }
    }

    pub fn tick(&mut self) -> bool {
        match self.state_stack.last_mut() {
            Some(State::Dispatch(dispatcher)) => {
                if let Some(quiz) = dispatcher.next() {
                    let progress = dispatcher.progress();
                    let state = QuestionState { quiz, progress };
                    self.state_stack.push(State::AskQuestion(state));
                } else {
                    self.state_stack.pop();
                }
                true
            }
            Some(_) => true,
            None => false,
        }
    }
}
