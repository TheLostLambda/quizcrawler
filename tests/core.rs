use insta::assert_ron_snapshot;
use quizcrawler::crawler::data::Crawler;
use std::fs;

const CONF_FILE: &str = "confs/borg.toml";

fn crawler() -> Crawler {
    let conf_str = fs::read_to_string(CONF_FILE).unwrap();
    Crawler::new(&conf_str).unwrap_or_else(|err| {
        panic!(
            "Failed to parse a config from string. The error was: {}\n The config string was: {}",
            err, conf_str
        );
    })
}

// It's a little bit gross to have crawler stuff in the core test file, but
// generating sections to test is messy without the crawler

#[test]
fn find_child_at_path() {
    let data_str = r#"
* Theme
** Topic 1
*** Subtopic A
  - Not quite right...
** Topic 2
*** Subtopic A
  - Well done!
"#;

    let path = vec!["Topic 2", "Subtopic A"];
    let section = &crawler().parse_sections(data_str)[0];
    let child = section.child_at_path(&path);
    assert_ron_snapshot!(child, {".**.last_correct" => "[last_correct]", ".**.id" => "[id]"});
}
