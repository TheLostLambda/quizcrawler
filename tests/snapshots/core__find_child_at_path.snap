---
source: tests/core.rs
expression: child
---
Some(Section(
  name: "Subtopic A",
  questions: [
    Question(
      data: Bullet(Bullet(
        body: "Well done!",
      )),
      comp_level: Trimmed,
      mastery: 0,
      correct: 0,
      seen: 0,
      last_correct: "[last_correct]",
    ),
  ],
  children: [],
))
